src/map.rs

```diff
@@ -956,6 +956,40 @@ where
             }
         }
     }
+    /// Retains only the elements specified by the predicate, and returns an iterator over the
+    /// removed items.
+    ///
+    /// In other words, move all pairs `(k, v)` such that `f(&k,&mut v)` returns `false` out
+    /// into another iterator.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use hashbrown::HashMap;
+    ///
+    /// let mut map: HashMap<i32, i32> = (0..8).map(|x|(x, x*10)).collect();
+    /// let drained = map.drain_filter(|&k, _| k % 2 == 0);
+    /// assert_eq!(drained.count(), 4);
+    /// assert_eq!(map.len(), 4);
+    /// ```
+    pub fn drain_filter<'a, F>(&'a mut self, mut f: F) -> impl Iterator<Item=(K,V)> + '_
+    where
+        F: 'a + FnMut(&K, &mut V) -> bool,
```

Comment:
Without the lifetime 'a I'm getting the error:
```
    Checking hashbrown v0.6.3 (/Users/julianknodt/Desktop/programming/projects/rustjects/hashbrown)
error[E0311]: the parameter type `F` may not live long enough
   --> src/map.rs:975:52
    |
975 |     pub fn drain_filter<F>(&mut self, mut f: F) -> impl Iterator<Item = (K, V)> + '_
    |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound for `F`
note: the parameter type `F` must be valid for the anonymous lifetime #1 defined on the method body at 975:5...
   --> src/map.rs:975:5
    |
975 | /     pub fn drain_filter<F>(&mut self, mut f: F) -> impl Iterator<Item = (K, V)> + '_
976 | |     where
977 | |         F: FnMut(&K, &mut V) -> bool,
978 | |     {
...   |
989 | |         }
990 | |     }
    | |_____^
note: ...so that the type `core::iter::FilterMap<raw::RawIter<(K, V)>, [closure@src/map.rs:980:42: 988:14 f:F, self:&mut map::HashMap<K, V, S>]>` will meet its required lifetime bounds
   --> src/map.rs:975:52
    |
975 |     pub fn drain_filter<F>(&mut self, mut f: F) -> impl Iterator<Item = (K, V)> + '_
    |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: could not compile `hashbrown`.

To learn more, run the command again with --verbose.
```

I think this will probably be stale if I implement a separate type for drain_filter though.